### PR TITLE
Create a callback logic inside the stream coordinator for when a stream ...

### PR DESCRIPTION
**Pull request by jguery**
...handler becomes the leader of a new stream/set of streams

### CODE REVIEW

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java [hunk 1]
```diff
@@ -157,8 +163,23 @@ public Long apply(StreamProperty property) {
...
-    return propertyStore.get().addChangeListener(streamName, new StreamPropertyChangeListener(streamAdmin,
-                                                                                              streamName, listener));
+    return propertyStore.get().addChangeListener(streamName,
+                                                 new StreamPropertyChangeListener(streamAdmin, streamName, listener));
+  }
+
+  @Override
+  public Cancellable addLeaderCallback(final StreamLeaderCallback callback) {
+    synchronized (this) {
+      leaderCallbacks.add(callback);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: Better not to add a user object to a set (as you don't know what contract they've implemented for equals/hashCode) for the canceling/removal purpose. It's better be a wrapped one.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/DistributedStreamCoordinator.java [hunk 2]
```diff
@@ -204,9 +206,16 @@ protected StreamsLeaderHandler() {
 
     @Override
     public void onChange(Collection<PartitionReplica> partitionReplicas) {
-      // TODO Use the names of the ParitionReplicas to retrieve the streams names and do something on them.
-      // TODO Here, we are in the master for all the streams represented by the partitions replicas. We need to do
-      // aggregation, etc.
+      Collection<String> streamNames = Collections2.transform(
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/DistributedStreamCoordinator.java]: Use `ImmutableSet.copyOf(Iterables.transform())` instead.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java [hunk 3]
```diff
@@ -168,6 +189,21 @@ protected final void shutDown() throws Exception {
   }
 
   /**
+   * Call all the callbacks that are interested in knowing that this coordinator is the leader of a set of Streams.
+   *
+   * @param streamNames set of Streams that this coordinator is the leader of
+   */
+  protected void callLeaderCallbacks(Set<String> streamNames) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: Call it `invokeLeaderCallbacks`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamLeaderCallback.java [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ */
+
+package co.cask.cdap.data.stream;
+
+import java.util.Set;
+
+/**
+ * Defines a behavior that is triggered when a Stream handler becomes leader of a Stream, or a collection of streams.
+ */
+public interface StreamLeaderCallback {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamLeaderCallback.java]: Better call it as `StreamLeaderListener`.

**Comment by chtyim** [comment]
Done review.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/DistributedStreamCoordinator.java (hunk 2 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/DistributedStreamCoordinator.java]: Ok.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java (hunk 3 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: ok.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamLeaderCallback.java (hunk 4 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamLeaderCallback.java]: Ok will change.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java (hunk 1 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: I see, will do.

**Comment by jguery** [comment]
Addressed your comments Terence, could you have another quick look please? Thanks

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinator.java [hunk 5]
```diff
@@ -65,7 +67,19 @@ public void setHandlerDiscoverable(Discoverable discoverable) {
...
-    // TODO implement the aggregation logic of the one stream handler process here
+    try {
+      callLeaderCallbacks();
+    } catch (Exception e) {
+      Throwables.propagate(e);
+    }
     return Futures.immediateFuture(null);
   }
+
+  private void callLeaderCallbacks() throws Exception {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinator.java]: `call` -> `invoke`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamCoordinator.java [hunk 6]
```diff
@@ -56,6 +56,17 @@
...
   /**
+   * This method is called every time the Stream handler in which this {@link StreamCoordinator}
+   * runs becomes the leader of a set of streams. Prior to this call, the Stream handler might
+   * already have been the leader of some of those streams.
+   *
+   * @param callback {@link StreamLeaderListener} called when this Stream handler becomes leader
+   *                 of a collection of streams
+   * @return A {@link Cancellable} to cancel the watch
+   */
+  Cancellable addLeaderCallback(StreamLeaderListener callback);
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamCoordinator.java]: Should be `addLeaderListener`.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinator.java (hunk 5 above)
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinator.java]: Also, callback -> listener.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java [hunk 7]
```diff
@@ -168,6 +197,21 @@ protected final void shutDown() throws Exception {
   }
 
   /**
+   * Call all the callbacks that are interested in knowing that this coordinator is the leader of a set of Streams.
+   *
+   * @param streamNames set of Streams that this coordinator is the leader of
+   */
+  protected void invokeLeaderCallbacks(Set<String> streamNames) {
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: callback -> listener.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java [hunk 8]
```diff
@@ -76,6 +80,8 @@ protected AbstractStreamCoordinator(StreamAdmin streamAdmin) {
 
     // Update action should be infrequent, hence just use an executor that create a new thread everytime.
     updateExecutor = ExecutorUtils.newThreadExecutor(Threads.createDaemonThreadFactory("stream-coordinator-update-%d"));
+
+    leaderCallbacks = Sets.newHashSet();
```
Comment (chtyim) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: callback -> listener.

**Comment by chtyim** [comment]
Just comments about unifying naming of `callback` to `listener`, `call` -> `invoke`.

**Comment by chtyim** [comment]
Otherwise LGTM.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinator.java (hunk 5 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/InMemoryStreamCoordinator.java]: Ok.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamCoordinator.java (hunk 6 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/StreamCoordinator.java]: Ok.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java (hunk 7 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: Ok.

File: cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java (hunk 8 above)
Comment (jguery) [inline cdap-data-fabric/src/main/java/co/cask/cdap/data/stream/AbstractStreamCoordinator.java]: Ok.

**Pull request merged**